lib/ts/recipe/session/sessionClass.ts

```diff
@@ -92,14 +102,52 @@ export default class Session implements SessionContainerInterface {
 
     // Any update to this function should also be reflected in the respective JWT version
     async mergeIntoAccessTokenPayload(accessTokenPayloadUpdate: any, userContext?: any): Promise<void> {
-        const updatedPayload = { ...this.getAccessTokenPayload(userContext), ...accessTokenPayloadUpdate };
+        const newAccessTokenPayload = { ...this.getAccessTokenPayload(userContext), ...accessTokenPayloadUpdate };
+        for (const key of protectedProps) {
+            delete newAccessTokenPayload[key];
+        }
+
         for (const key of Object.keys(accessTokenPayloadUpdate)) {
             if (accessTokenPayloadUpdate[key] === null) {
-                delete updatedPayload[key];
+                delete newAccessTokenPayload[key];
             }
         }
 
-        await this.updateAccessTokenPayload(updatedPayload, userContext);
+        let response = await this.helpers.getRecipeImpl().regenerateAccessToken({
+            accessToken: this.getAccessToken(),
+            newAccessTokenPayload,
+            userContext: userContext === undefined ? {} : userContext,
+        });
+
+        if (response === undefined) {
+            throw new STError({
+                message: "Session does not exist anymore",
+                type: STError.UNAUTHORISED,
+            });
+        }
+
+        this.userDataInAccessToken = response.session.userDataInJWT;
+        if (response.accessToken !== undefined) {
+            this.accessToken = response.accessToken.token;
```

Comment:
- if exposeToFrontend is true, then we also need to set the access token in header? Where else is this missed?
- consider moving this part of setting tokens to a single place which checks for the new config (exposeToFrontend), and sets the tokens accordingly